modules/siddhi-core/src/test/java/org/wso2/siddhi/core/query/pattern/WithinPatternTestCase.java

```diff
@@ -209,5 +209,181 @@ public void receive(long timestamp, Event[] inEvents, Event[] removeEvents) {
         siddhiAppRuntime.shutdown();
     }
 
+    @Test
+    public void testQuery4() throws InterruptedException {
+        log.info("testPatternWithin4 - OUT 1 : Within clause for grouped pattern states");
+
+        SiddhiManager siddhiManager = new SiddhiManager();
+
+        String streams = "" +
+                "define stream Stream1 (symbol string, price float, volume int); ";
+
+        String query = "" +
+                "@info(name = 'query1') " +
+                "from every (e1=Stream1 -> e2=Stream1[symbol == e1.symbol]) within 5 sec " +
+                "select e1.symbol as symbol1, e1.volume as volume1, e2.symbol as symbol2, e2.volume as volume2 " +
+                "insert into OutputStream ;";
+
+        SiddhiAppRuntime siddhiAppRuntime = siddhiManager.createSiddhiAppRuntime(streams + query);
+
+        siddhiAppRuntime.addCallback("query1", new QueryCallback() {
+            @Override
+            public void receive(long timestamp, Event[] inEvents, Event[] removeEvents) {
+                EventPrinter.print(timestamp, inEvents, removeEvents);
+                if (inEvents != null) {
+                    inEventCount = inEventCount + inEvents.length;
+                    AssertJUnit.assertArrayEquals(new Object[]{"WSO2", 200, "WSO2", 250}, inEvents[0].getData());
```

Comment:
Shouldn't the output need to be {"WSO2", 150, "WSO2", 200} ??

Reply:
I believe it will be like this as at the 6000 ms delay the pattern should rest.

Reply:
The thing is the way how expiry handled is a bit tricky when it comes to grouped states. Seems we need some refactoring on this to get such a behavior... Let's have a chat. I'll check further on possible options.